File: gprolog-1.3.0/interpreter/testenv.pl

testOk :- 
 catch( teststate( CountOk, CountFail, CountLatest), error(existence_error(_,_),_), 
  ( CountOk= 0, CountFail= 0, CountLatest= 0)
 ),
 retractall( teststate( _, _, _)),
 CountOk_ is 1 + CountOk,
 assertz( teststate( CountOk_, CountFail, CountLatest)),
 true.

testFail :- 
 catch( teststate( CountOk, CountFail, CountLatest), error(existence_error(_,_),_), 
  ( CountOk= 0, CountFail= 0, CountLatest= 0)
 ),
 retractall( teststate( _, _, _)),
 CountFail_ is 1 + CountFail,
 assertz( teststate( CountOk, CountFail_, CountLatest)),
 format( "test failed~n", []),
 true.

checkComplainAndAdjustExpected( Count) :-

 catch( teststate( CountOk, CountFail, CountLatest), error(existence_error(_,_),_), 
  ( CountOk= 0, CountFail= 0, CountLatest= 0)
 ),

 CountTests is CountOk + CountFail,
 CountTestsLocal is CountTests - CountLatest,

 retractall( teststate( _, _, _)),
 assertz( teststate( CountOk, CountFail, CountTests)),

 (
  Count = CountTestsLocal -> true
  ;
  format( "Expected Test Count (from last checkpoint to here) doesn't match: countTestsLocal: ~d cound: ~d~n", [ CountTestsLocal, Count])
 ),

 true.

result( OUT) :-
 %open_output_chars_stream( S),
 open_output_codes_stream( S),
 catch( teststate( CountOk, CountFail, CountLatest), error(existence_error(_,_),_), 
  ( CountOk= 0, CountFail= 0, CountLatest= 0)
 ),
 format( S, "tests: ok: ~d fail: ~d countLatest: ~d~n", [ CountOk, CountFail, CountLatest]),
 %close_output_chars_stream( S, OUT),
 close_output_codes_stream( S, OUT),
 true.

File: gforth-0.7.0/interpreter/testenv.fs
create tst_countOk 
create tst_countFail 
create tst_countLatest 

\ boolean : -1 true, 0 false

: tst_test 
 if
  tst_countOk @ 1 + tst_countOk !
 else
  tst_countFail @ 1 + tst_countFail !
  ." test failed"
  cr
 endif
;

: tst_checkComplainAndAdjustExpected
 \ count
 tst_countOk @ tst_countFail @ + \ count countTests
 dup tst_countLatest @ - \ count countTests countTestsLocal
 swap \ count countTestsLocal countTests
 tst_countLatest ! \ count countTestsLocal
 2dup \ count countTestsLocal count countTestsLocal
 = \ count countTestsLocal (= count countTestsLocal)
 
 if
  -1 \ count countTestsLocal ret
 else
  2dup \ count countTestsLocal count countTestsLocal
  ." Expected Test Count (from last checkpoint to here) doesn't match: "
  ." countTestsLocal: "
  .
  ." count: "
  .
  cr
  0 \ count countTestsLocal ret
 endif
 
 rot drop drop \ ret
;

: tst_result
 ." tests: "
 ." ok: "
 tst_countOk @ .
 ." fail: "
 tst_countFail @ .
 ." countLatest: "
 tst_countLatest @ .
 cr
;

0 tst_countOk !
0 tst_countFail !
0 tst_countLatest !

